tests/regression/test_linesmoother.py

```diff
@@ -0,0 +1,102 @@
+import pytest
+from firedrake import *
+
+
+@pytest.fixture(params=["Interval", "Triangle", "Quad"])
+def mesh_type(request):
+    return request.param
+
+
+@pytest.fixture
+def mesh(mesh_type):
+    if mesh_type == "Interval":
+        return ExtrudedMesh(IntervalMesh(10, 100), 5, 1/5)
+    elif mesh_type == "Triangle":
+        return ExtrudedMesh(RectangleMesh(10, 10, 100, 100), 5, 1/5)
+    elif mesh_type == "Quad":
+        return ExtrudedMesh(RectangleMesh(10, 10, 100, 100,
+                                          quadrilateral=True), 5, 1/5)
+
+
+@pytest.fixture
+def S1family(mesh_type):
+    if mesh_type == "Interval":
+        return "CG"
+    elif mesh_type == "Triangle":
+        return "BDM"
+    elif mesh_type == "Quad":
+        return "RTCF"
+
+
+@pytest.fixture
+def expected(mesh_type):
+    if mesh_type == "Interval":
+        return [8, 13]
+    elif mesh_type == "Triangle":
+        return [10, 26]
+    elif mesh_type == "Quad":
+        return [9, 20]
+
+
+def test_linesmoother(mesh, S1family, expected):
+
+    nits = []
+    for degree in range(2):
+        S1 = FiniteElement(S1family, mesh._base_mesh.ufl_cell(), degree+1)
+        S2 = FiniteElement("DG", mesh._base_mesh.ufl_cell(), degree)
+        T0 = FiniteElement("CG", interval, degree+1)
+        T1 = FiniteElement("DG", interval, degree)
+
+        V2h_elt = HDiv(TensorProductElement(S1, T1))
+        V2t_elt = TensorProductElement(S2, T0)
+        V2v_elt = HDiv(V2t_elt)
+        V3_elt = TensorProductElement(S2, T1)
+        V2_elt = V2h_elt + V2v_elt
+
+        V = FunctionSpace(mesh, V2_elt)
+        Q = FunctionSpace(mesh, V3_elt)
+
+        W = MixedFunctionSpace((V, Q))
+
+        u, p = TrialFunctions(W)
+        v, q = TestFunctions(W)
+
+        a = (inner(v, u) - div(v)*p + p*q + div(u)*q)*dx
+        bcs = [DirichletBC(W.sub(0), 0, "on_boundary"),
+               DirichletBC(W.sub(0), 0, "top"),
+               DirichletBC(W.sub(0), 0, "bottom")]
+
+        x = SpatialCoordinate(mesh)
+        if len(x) == 2:
+            rsq = (x[0]-50)**2/20**2 + (x[1]-0.5)**2/0.2**2
+        else:
+            rsq = (x[0]-50)**2/20**2 + (x[1] - 50)**2/20**2 + (x[2]-0.5)**2/0.2**2
+        f = exp(-rsq)
+
+        L = q*f*dx
```

Comment:
The part of me that is wanting to merge the complex branch asks you to write `inner(trial, test)` and `inner(coefficient, test)` everywhere

Reply:
Hopefully I've done that now